### PR TITLE
Make SearchForDockerfile return DockerfileImage

### PR DESCRIPTION
**Pull request by tkdchen**
This change makes it easier to access other properties of an
ImageComponent.

Signed-off-by: Chenxiong Qi <cqi@redhat.com>

### CODE REVIEW

File: pkg/devfile/detect.go [hunk 1]
```diff
@@ -196,31 +202,28 @@ func AnalyzePath(a Alizer, localpath, context, devfileRegistryURL string, devfil
...
 		// Only check for the Dockerfile Uri at this point, in later stages we need to account for Dockerfile from Git & the Registry
 		if component.Image != nil && component.Image.Dockerfile != nil && component.Image.Dockerfile.DockerfileSrc.Uri != "" {
-			dockerfile = component.Image.Dockerfile.DockerfileSrc.Uri
-			break
+			return component.Image.Dockerfile, nil
 		}
 	}
 
-	return dockerfile, nil
+	return nil, nil
```
Comment (Michkov) [inline pkg/devfile/detect.go:226]: you can return empty object, so you will not have issue with nil pointer -> `return &v1alpha2.DockerfileImage{}, nil`

File: pkg/devfile/detect_test.go [hunk 2]
```diff
@@ -144,13 +144,13 @@ components:
...
 			devfileBytes := []byte(tt.devfileString)
-			gotUri, err := SearchForDockerfile(devfileBytes)
+			dockerfileImage, err := SearchForDockerfile(devfileBytes)
 			if !tt.wantErr && err != nil {
 				t.Errorf("Unexpected err: %+v", err)
 			} else if tt.wantErr && err == nil {
 				t.Errorf("Expected error but got nil")
-			} else if gotUri != tt.wantUri {
-				t.Errorf("Expected %v but got %v", tt.wantUri, gotUri)
+			} else if dockerfileImage.Uri != tt.wantUri {
```
Comment (Michkov) [inline pkg/devfile/detect_test.go:152]: test failed here since dockerfileImage is nil

**Review by gabemontero — APPROVED** [review]
per our slack discussions +1 on using a pointer to deal with existence vs. default/empty

feel like the codecov complaints are a bit unfair :-)  ... so of the error paths it was complaining about unit test coverage for were there previously; this PR did not add those error paths

placing a 

/hold 

just in case my approve adds an lgtm for this PR; want to make sure @mmorhun has a chance to review

**Comment by gabemontero** [comment]
ok the PR was not already approved, so I can remove the hold and this won't merge with my lgtm

/hold cancel

**Comment by tkdchen** [comment]
A new update is pushed to this PR. The update is in SearchForDockerfile function, which does not parse and validate an empty devfile and just return nil immediately as non-existing dockerfile. Tests are updated accordingly.

**Review by Michkov — APPROVED** [review]

**Review by jduimovich — APPROVED** [review]
LGTM

**Pull request merged**